net: pkt: context: Respect reserved bytes

rettichschnidi opened the pull request:
When operating on a net_pkt using the curser functionality, reserved
bytes (headroom) must be respected.

This commit achieves this by updating the size member of the
net_buf_simple structure when reserving (and pushing) bytes.

While this solution changes the semantic of the size member, it requires
a low amount of code changes in addition to allowing the pattern of
writing (up to) size bytes starting at the data pointer to work properly
even in the presence of reserved bytes.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

jukkar commented on include/net/buf.h:
```diff
@@ -91,7 +91,7 @@ struct net_buf_simple {
 	/** Length of the data behind the data pointer. */
```
typo in the commit msg, s/curser/cursor/
Also you can shorten the commit subject as the "context" is not really needed here, just write it
`net: pkt: Respect reserved bytes`

rettichschnidi replied:
fixed

jukkar reviewed (commented):
This looks reasonable.
@jhedberg does this cause issues in Bluetooth side?

jhedberg commented:
I need to look more carefully at this. Meanwhile, please split the net_buf changes into a separate commit with a `net: buf: ...` prefix. The net_buf semantics and API terminology has been created by using Linux sk_buff as reference - I think we should keep the same semantics in order not to confuse people. The net_buf API is not a 1:1 mapping to sk_buff, so I'll need to check if this is true for the parts that this PR affects.

rettichschnidi committed:
net: pkt: Expose headroom edge case

When operating on a net_pkt using the cursor functionality, reserved
bytes (headroom) are currently not respected.

This commit adds test cases exposing this problem. Fix will be
introduced in a later commit.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi commented:
Just reworked the PR.

I guess all of this boils down to two questions:
* Is it desirable to use the net_pkt cursor as done in the first commit? (I think so)
* Is the solution proposed in the 2nd commit the right/best one? (Unsure about this)

jhedberg commented:
>  In order to (safely) append data to a net_buf object, the user must query net_buf_headroom() to ensure enough space is left.

@rettichschnidi I don't understand this comment. To check if there's space to append data you only need to query `net_buf_tailroom()`

jhedberg commented:
@rettichschnidi I get the feeling that the background of this PR is the fact that you've been using `buf->size` as a primary API, when it was mainly designed to be useful only for net_buf internal purposes. From the net_buf API user perspective the `net_buf_headroom()` and `net_buf_tailroom()` are the primary methods that are expected to be used in order to determine if there's space to append or prepend more data to the buffer.

rettichschnidi committed:
net: buf: Simplify querying maximum len

net_buf_max_len() provides the maximum number of bytes which can be put
behind the data pointer. This provides a save alternative to using the
size field of the net_buf structure directly, which does not take the
reserved bytes (headroom) into account.

This commit also replaces the usage of the size field in places where
size got used directly. Code has not been adjusted when it is easy to
recognise that the buffer does not have any reserved bytes, which is the
case after allocation or reset. Same goes for the faulty usage by
net_pkt as exposed by the last commit and begin fixed by a separate
commit.

Even though it would be cleaner, I decided to not rename the size field
to e.g. __buf_size in order to keep the amount of code changes low.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
net: pkt: Fix headroom edge case

This commit fixes some edge cases when using net_bufs with reserved
bytes (headroom) as fragments of a net_pkt.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi commented:
> @rettichschnidi I don't understand this comment. To check if there's space to append data you only need to query `net_buf_tailroom()`

My comment is misleading. I will clarify this with an updated comment.

> @rettichschnidi I get the feeling that the background of this PR is the fact that you've been using `buf->size` as a primary API, when it was mainly designed to be useful only for net_buf internal purposes. From the net_buf API user perspective the `net_buf_headroom()` and `net_buf_tailroom()` are the primary methods that are expected to be used in order to determine if there's space to append or prepend more data to the buffer.

Yes, the problem stems from the (ab)use of `buf->size`. However, it is not (just) me, but also existing code in the Zephyr code base, like the code implementing the net_pkt_cursor manipulations.

I will come up with an updated proposal which does not not change the (the semantic of the) size member of net_buf, but using it properly instead.

rettichschnidi commented:
I am unsure why buildkite fails - neither `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T tests/kernel/lifo/lifo_usage/` nor `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T samples/kernel/condition_variables/condvar/` errors on my own machine.

jukkar commented:
> I am unsure why buildkite fails - neither `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T tests/kernel/lifo/lifo_usage/` nor `scripts/twister --inline-logs -N -v --integration --retry-failed 3 -T samples/kernel/condition_variables/condvar/` errors on my own machine.

There seems to be some unrelated issues in the kernel side and some tests are indeed failing. Dunno what is wrong there.
But there is one net_pkt failure here
```
/workdir/zephyr/subsys/net/ip/net_pkt.c: In function 'net_pkt_print_frags':
/workdir/zephyr/subsys/net/ip/net_pkt.c:347:45: error: 'buf' undeclared (first use in this function)
  347 |     count, frag, frag->len, net_buf_max_len(buf),
      |                                             ^~~
```
